src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+﻿// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace MongoDB.Analyzer.Core.Builders;
+
+internal static class BuildersResolveVariables
+{
+    private sealed record VariableValue(
+        ExpressionAnalysisContext ExpressionAnalysisContext,
+        List<Location> Locations);
+
+    private sealed record ContextInformation(
+        SemanticModel SemanticModel,
+        List<VariableValue> VariableValuesList,
+        Dictionary<string, VariableValue> VariableValues,
+        HashSet<SyntaxNode> BuilderNodesProcessed,
```

Comment:
Maybe more descriprtive naming? Try thinking about something like

VariableValues -> CurrentVariablesValues and 
VariableValuesList -> ProcessedVariables

Reply:
Done